File: ui-anim-word-lwp/src/main/java/com/sfaxdoird/anim/word/AnimWord2dWallpaper.java
package com.sfaxdoird.anim.word;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.os.Handler;
import android.service.wallpaper.WallpaperService;
import android.view.MotionEvent;
import android.view.SurfaceHolder;

import com.sfaxdroid.base.SharedPrefsUtils;

import java.io.File;

public class AnimWord2dWallpaper extends WallpaperService {

    private static int currentPhoto = 1;
    private Paint mPaintOption;

    @Override
    public Engine onCreateEngine() {
        SharedPrefsUtils pref = new SharedPrefsUtils(this);
        int color = pref.GetSetting("DouaLwpColor", -4522170);
        int textSize = pref.GetSetting("nameofallahtextsize", 1) * 20;
        int fontStyle = pref.GetSetting("nameofallahfontstyle", 1);
        return new IslamicEngine(color, textSize, fontStyle);
    }

    class IslamicEngine extends Engine {
        private final Handler mHandler = new Handler();
        private final SurfaceHolder mHolder = getSurfaceHolder();
        private File mFilesDir = getExternalFilesDir("");
        private int mScreenHeight;
        private int mScreenWidth;
        private Bitmap mBackground;
        private boolean mVisible;
        private final Runnable mDrawPattern = new Runnable() {
            public void run() {
                drawFrame();
            }
        };
        private BitmapFactory.Options mOptions = new BitmapFactory.Options();
        private int mColor;

        IslamicEngine(int color, int textSize, int fontStyle) {
            mColor = color;
            final Paint mPaint = new Paint();
            mPaint.setColor(0xffffffff);
            mPaint.setAntiAlias(true);
            mPaint.setStrokeWidth(2);
            mPaint.setStrokeCap(Paint.Cap.ROUND);
            mPaint.setStyle(Paint.Style.STROKE);
            mOptions.inPreferredConfig = Bitmap.Config.ARGB_8888;

            mPaintOption = new Paint();
            mPaintOption.setColor(mColor);

            mPaintOption.setTextSize(textSize * getResources().getDisplayMetrics().density);
            if (getTypeFace(fontStyle) != null)
                mPaintOption.setTypeface(getTypeFace(fontStyle));
        }


        Typeface getTypeFace(int mTypefaceNum) {
            try {
                if (mTypefaceNum == 1)
                    return Typeface.createFromAsset(getAssets(), "arabicfont1.otf");
                if (mTypefaceNum == 2)
                    return Typeface.createFromAsset(getAssets(), "arabicfont2.ttf");
                else if (mTypefaceNum == 3)
                    return Typeface.createFromAsset(getAssets(), "arabicfont3.ttf");
                else if (mTypefaceNum == 4)
                    return Typeface.createFromAsset(getAssets(), "arabicfont4.otf");
                else if (mTypefaceNum == 5)
                    return Typeface.createFromAsset(getAssets(), "arabicfont5.ttf");
                else if (mTypefaceNum == 6)
                    return Typeface.createFromAsset(getAssets(), "arabicfont6.ttf");
                else if (mTypefaceNum == 7)
                    return Typeface.createFromAsset(getAssets(), "arabicfont7.ttf");
                else
                    return Typeface.createFromAsset(getAssets(), "arabicfont8.ttf");
            } catch (Exception e) {
                return Typeface.DEFAULT;
            }
        }


        @Override
        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
            setTouchEventsEnabled(true);
        }

        @Override
        public void onDestroy() {
            super.onDestroy();
            mHandler.removeCallbacks(mDrawPattern);
        }

        @Override
        public void onVisibilityChanged(boolean visible) {
            mVisible = visible;
            if (visible) {
                drawFrame();
            } else {
                mHandler.removeCallbacks(mDrawPattern);
            }
        }

        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format,
                                     int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);
            if (mBackground != null) {
                mBackground.recycle();
                mBackground = null;
            }
            Bitmap mBtm = BitmapFactory.decodeFile(mFilesDir + "/" + com.sfaxdroid.base.Constants.DOUA_PNG_BACKFROUND_FILE_NAME,
                    mOptions);
            if (mBtm != null) {
                mBackground = Bitmap.createScaledBitmap(mBtm, width, height, true);
                mBtm.recycle();
            }
            if (mBackground != null)
                drawFrame();
        }

        @Override
        public void onSurfaceCreated(SurfaceHolder holder) {
            super.onSurfaceCreated(holder);
        }

        @Override
        public void onSurfaceDestroyed(SurfaceHolder holder) {
            super.onSurfaceDestroyed(holder);
            mVisible = false;
            mHandler.removeCallbacks(mDrawPattern);
            if (mBackground != null)
                mBackground.recycle();
        }

        @Override
        public void onOffsetsChanged(float xOffset, float yOffset, float xStep,
                                     float yStep, int xPixels, int yPixels) {
            drawFrame();
        }

        @Override
        public void onTouchEvent(MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
                com.sfaxdroid.base.Constants.nbIncrementationAfterChange = com.sfaxdroid.base.Constants.nbIncrementationAfterChange + 1;
                if (mScreenWidth > 0
                        && com.sfaxdroid.base.Constants.ifBackgroundChanged
                        && com.sfaxdroid.base.Constants.nbIncrementationAfterChange == 5) {
                    com.sfaxdroid.base.Constants.nbIncrementationAfterChange = 0;
                    com.sfaxdroid.base.Constants.ifBackgroundChanged = false;
                    mBackground = Bitmap.createScaledBitmap(BitmapFactory.decodeFile(mFilesDir + "/"
                                    + com.sfaxdroid.base.Constants.DOUA_PNG_BACKFROUND_FILE_NAME, mOptions),
                            mScreenWidth, mScreenHeight, true);
                }

                //todo maybe change color
                mPaintOption.setColor(mColor);

                if (currentPhoto == 98) {
                    currentPhoto = 0;
                }
                currentPhoto++;
            }
            super.onTouchEvent(event);
        }

        void drawFrame() {
            Canvas canvas = null;
            try {
                canvas = mHolder.lockCanvas();
                if (canvas != null) {
                    drawPattern(canvas);
                }
            } finally {
                if (canvas != null)
                    mHolder.unlockCanvasAndPost(canvas);
            }
            mHandler.removeCallbacks(mDrawPattern);
            if (mVisible) {
                mHandler.postDelayed(mDrawPattern, 1000 / 25);
            }
        }

        void drawPattern(Canvas c) {
            if (mBackground != null && mBackground.getWidth() != 0) {
                c.save();
                c.drawColor(0xff000000);
                mScreenWidth = c.getWidth();
                mScreenHeight = c.getHeight();
                if (!mBackground.isRecycled()) {
                    c.drawBitmap(mBackground, 0, 0, null);
                }
                String header = Constants.NAME_OF_ALLAH_TAB.get(currentPhoto);
                Rect r = new Rect();
                c.getClipBounds(r);
                int cHeight = r.height();
                int cWidth = r.width();
                mPaintOption.setTextAlign(Paint.Align.LEFT);
                mPaintOption.getTextBounds(header, 0, header.length(), r);
                float x = cWidth / 2f - r.width() / 2f - r.left;
                float y = cHeight / 2f + r.height() / 2f - r.bottom;
                c.drawText(header, x, y, mPaintOption);
                c.restore();
            }
        }
    }
}

File: ui-anim-word-lwp/src/main/java/com/sfaxdoird/anim/word/AnimWord2dWallpaper.kt
package com.sfaxdoird.anim.word

import android.graphics.*
import android.os.Handler
import android.service.wallpaper.WallpaperService
import android.view.MotionEvent
import android.view.SurfaceHolder
import com.sfaxdroid.base.SharedPrefsUtils


class AnimWord2dWallpaper : WallpaperService() {
    override fun onCreateEngine(): Engine {
        val pref = SharedPrefsUtils(this)
        val color = pref.GetSetting(com.sfaxdroid.bases.Constants.WALLPAPER_COLOR, -4522170)
        val textSize = pref.GetSetting("2dwallpapertextsize", 1) * 20
        val fontStyle = pref.GetSetting("2dwallpaperfontstyle", 1)
        return IslamicEngine(color, textSize, fontStyle)
    }

    internal inner class IslamicEngine(private val mColor: Int, textSize: Int, fontStyle: Int) :
        Engine() {

        private var paintOption: Paint
        private val handler = Handler()
        private val mHolder = surfaceHolder
        private val filesDir = getExternalFilesDir("")
        private var screenHeight = 0
        private var screenWidth = 0
        private var background: Bitmap? = null
        private var mVisible = false
        private val drawPattern = Runnable { drawFrame() }
        private val options = BitmapFactory.Options()

        init {
            options.inPreferredConfig = Bitmap.Config.ARGB_8888
            paintOption = Paint().apply { }
            paintOption.color = mColor
            paintOption.textSize = textSize * resources.displayMetrics.density
            getTypeFace(fontStyle)?.let {
                paintOption.typeface = it
            }
        }

        private fun getTypeFace(mTypefaceNum: Int): Typeface? {
            return try {
                Typeface.createFromAsset(
                    assets,
                    "arabicfont$mTypefaceNum.ttf"
                )
            } catch (e: Exception) {
                return try {
                    Typeface.createFromAsset(
                        assets,
                        "arabicfont$mTypefaceNum.otf"
                    )
                } catch (e: Exception) {
                    Typeface.DEFAULT
                }
            }
        }

        override fun onCreate(surfaceHolder: SurfaceHolder) {
            super.onCreate(surfaceHolder)
            setTouchEventsEnabled(true)
        }

        override fun onDestroy() {
            super.onDestroy()
            handler.removeCallbacks(drawPattern)
        }

        override fun onVisibilityChanged(visible: Boolean) {
            mVisible = visible
            if (visible) {
                drawFrame()
            } else {
                handler.removeCallbacks(drawPattern)
            }
        }

        override fun onSurfaceChanged(
            holder: SurfaceHolder, format: Int,
            width: Int, height: Int
        ) {
            super.onSurfaceChanged(holder, format, width, height)

            background?.let {
                it.recycle()
                background = null
            }

            BitmapFactory.decodeFile(
                filesDir.toString() + "/" + com.sfaxdroid.base.Constants.PNG_BACKFROUND_FILE_NAME,
                options
            )?.also {
                background = Bitmap.createScaledBitmap(it, width, height, true)
                it.recycle()
            }
        }

        override fun onSurfaceDestroyed(holder: SurfaceHolder) {
            super.onSurfaceDestroyed(holder)
            mVisible = false
            handler.removeCallbacks(drawPattern)
            background?.recycle()
        }

        override fun onOffsetsChanged(
            xOffset: Float, yOffset: Float, xStep: Float,
            yStep: Float, xPixels: Int, yPixels: Int
        ) {
            drawFrame()
        }

        override fun onTouchEvent(event: MotionEvent) {
            if (event.action == MotionEvent.ACTION_DOWN) {
                com.sfaxdroid.base.Constants.nbIncrementationAfterChange =
                    com.sfaxdroid.base.Constants.nbIncrementationAfterChange + 1
                if (screenWidth > 0 && com.sfaxdroid.base.Constants.ifBackgroundChanged
                    && com.sfaxdroid.base.Constants.nbIncrementationAfterChange == 5
                ) {
                    com.sfaxdroid.base.Constants.nbIncrementationAfterChange = 0
                    com.sfaxdroid.base.Constants.ifBackgroundChanged = false
                    background = Bitmap.createScaledBitmap(
                        BitmapFactory.decodeFile(
                            filesDir.toString() + "/"
                                    + com.sfaxdroid.base.Constants.PNG_BACKFROUND_FILE_NAME,
                            options
                        ),
                        screenWidth, screenHeight, true
                    )
                }

                //todo maybe change color
                paintOption.color = mColor
                if (currentPhoto == 98) {
                    currentPhoto = 0
                }
                currentPhoto++
            }
            super.onTouchEvent(event)
        }

        private fun drawFrame() {
            var canvas: Canvas? = null
            try {
                canvas = mHolder.lockCanvas()
                canvas?.let { drawPattern(it) }
            } finally {
                canvas?.let {
                    mHolder.unlockCanvasAndPost(canvas)
                }
            }
            handler.removeCallbacks(drawPattern)
            if (mVisible) {
                handler.postDelayed(drawPattern, 1000 / 25.toLong())
            }
        }

        private fun drawWallpaper(canvas: Canvas) {
            canvas.save()
            canvas.drawColor(-0x1000000)
            background?.let {
                canvas.drawBitmap(it, 0f, 0f, null)
            }
        }

        private fun draw2dImg(canvas: Canvas) {

            val header =
                Constants.NAME_OF_ALLAH_TAB[currentPhoto]

            val bounds = Rect()
            paintOption.getTextBounds(header, 0, header.length, bounds)
            val x = canvas.width / 2 - bounds.width() / 2
            val y = canvas.height / 2 - bounds.height() / 2

            canvas.drawText(header, x.toFloat(), y.toFloat(), paintOption)
        }

        private fun drawPattern(canvas: Canvas) {
            screenWidth = canvas.width
            screenHeight = canvas.height
            drawWallpaper(canvas)
            draw2dImg(canvas)
            canvas.restore()
        }
    }

    companion object {
        private var currentPhoto = 1
    }
}